File: 2571-minimum-operations-to-reduce-an-integer-to-0/2571-minimum-operations-to-reduce-an-integer-to-0.cpp
class Solution {
    int power2(int n){
        n |= n >> 1;
        n |= n >> 2;
        n |= n >> 4;
        n |= n >> 8;
        n |= n >> 16;
        return n ^ (n>>1);
    }
public:
    int minOperations(int n) {
        if( (n & (n - 1)) == 0)
            return 1;
        int low = power2(n), high = low * 2;
        return 1  + minOperations(min(n -low, high - n));
    }
};

File: 2571-minimum-operations-to-reduce-an-integer-to-0/2571-minimum-operations-to-reduce-an-integer-to-0.py
class Solution:
    def minOperations(self, n: int) -> int:
        if n & n -1 == 0:
            return 1
        temp = n
        n = n | n >> 1
        n = n | n >> 2
        n = n | n >> 4
        n = n | n >> 8
        n = n | n >> 16
        n = n ^ (n >> 1)
        low = n 
        high = 2 * n
        temp = min(high - temp, temp - low)
        S = Solution()
        return 1 + S.minOperations(temp)
        
        